lib/ocpp/v201/charge_point.cpp

```diff
@@ -3296,6 +3393,132 @@ void ChargePoint::handle_heartbeat_response(CallResult<HeartbeatResponse> call)
     }
 }
 
+void ChargePoint::handle_reserve_now_request(Call<ReserveNowRequest> call) {
+    ReserveNowResponse response;
+    response.status = ReserveNowStatusEnum::Rejected;
+    if (!this->callbacks.reserve_now_callback.has_value() ||
+        !this->device_model->get_optional_value<bool>(ControllerComponentVariables::ReservationCtrlrAvailable)
+             .value_or(false) ||
+        !this->device_model->get_optional_value<bool>(ControllerComponentVariables::ReservationCtrlrEnabled)
+             .value_or(false)) {
+        // Reservation not available / implemented, return 'Rejected'.
+        // H01.FR.01
+        EVLOG_info << "Receiving a reservation request, but reservation is not implemented.";
+        const ocpp::CallResult<ReserveNowResponse> call_result(response, call.uniqueId);
+        this->send<ReserveNowResponse>(call_result);
+        return;
+    }
+
+    // Check if we need a specific evse id during a reservation and if that is the case, if we recevied an evse id.
+    const ReserveNowRequest request = call.msg;
+    if (!request.evseId.has_value() &&
+        !this->device_model->get_optional_value<bool>(ControllerComponentVariables::ReservationCtrlrNonEvseSpecific)
+             .value_or(false)) {
+        // H01.FR.19
+        EVLOG_warning
+            << "Trying to make a reservation, but no evse id was given while it should be sent in the request.";
+        const ocpp::CallResult<ReserveNowResponse> call_result(response, call.uniqueId);
+        this->send<ReserveNowResponse>(call_result);
+        return;
+    }
+
+    const std::optional<int32_t> evse_id = request.evseId;
+
+    ConnectorStatusEnum connector_status = ConnectorStatusEnum::Unavailable;
+
+    if (evse_id.has_value()) {
+        if (evse_id.value() < 0 || !evse_manager->does_evse_exist(evse_id.value())) {
```

Comment:
I see I missed some checks in the evse manager. The smaller than 0 check should definitely be there (also in the `get_evse` functions. Then `evse_id.value() < 0` would not be necessary anymore